Here is the page that has list of all available bert models on tensorflow hub that one can download and make use of.

https://tfhub.dev/google/collections/bert/1

Here is the information on basic uncased BERT model,

https://tfhub.dev/tensorflow/bert_en_uncased_L-12_H-768_A-12/4

It uses L=12 hidden layers (i.e., Transformer blocks), a hidden size of H=768, and A=12 attention heads. This model has been pre-trained for English on the Wikipedia and BooksCorpus.

In [3]:
import tensorflow as tf
import tensorflow_hub as hub
import tensorflow_text as text

In [4]:
preprocess_url = "https://tfhub.dev/tensorflow/bert_en_uncased_preprocess/3"
encoder_url = "https://tfhub.dev/tensorflow/bert_en_uncased_L-12_H-768_A-12/4"

In [5]:
bert_preprocess_model = hub.KerasLayer(preprocess_url)

In [8]:
text_test = ["this is such an amazing movie!","The movie was great", "The movie was meh.", "The movie was okish.", "The movie was terrible..."]
text_preprocessed = bert_preprocess_model(text_test)
text_preprocessed.keys()

dict_keys(['input_type_ids', 'input_mask', 'input_word_ids'])

In [9]:
text_preprocessed['input_word_ids']

# CLS The movie was great SEP

<tf.Tensor: shape=(5, 128), dtype=int32, numpy=
array([[ 101, 2023, 2003, 2107, 2019, 6429, 3185,  999,  102,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
           0,    0,    0,    0,    0,    0,    0],
       [ 101, 1996, 3185, 2001, 2307,  102,    0,    0,    0,    0,    0,
           0,

101 --> CLS token

102 --> SEP token

BERT uses CLS as a special token at the begining of each setence whereas SEP as a special token to separate two sentences or end sinle sentece

In [13]:
berd_model = hub.KerasLayer(encoder_url)

In [14]:
bert_results = berd_model(text_preprocessed)
bert_results.keys()

dict_keys(['pooled_output', 'sequence_output', 'encoder_outputs', 'default'])

In [16]:
bert_results['encoder_outputs']

[<tf.Tensor: shape=(5, 128, 768), dtype=float32, numpy=
 array([[[ 0.16973737, -0.00696089, -0.06277221, ...,  0.08287903,
           0.00515024,  0.05787851],
         [-0.28907266,  0.4814251 ,  0.24022973, ..., -0.0871459 ,
           0.7371136 ,  0.04800112],
         [-0.7893274 , -0.23629351, -0.23939705, ...,  0.05862369,
           0.55842096,  0.42949197],
         ...,
         [-0.07927103, -0.23740894,  0.66923314, ...,  0.44114354,
          -0.18840502,  0.05491538],
         [-0.15417233, -0.24830565,  0.54970765, ...,  0.45950484,
          -0.10652437, -0.03091783],
         [-0.04351531, -0.19316751,  0.58315086, ...,  0.7488989 ,
          -0.3631808 , -0.07057214]],
 
        [[ 0.20184705,  0.08458187, -0.13259672, ...,  0.06339414,
           0.06640753,  0.11473774],
         [ 0.0527114 ,  0.25894752, -0.08798096, ...,  0.24258612,
           0.6707544 , -0.99705905],
         [-0.25571465,  0.5477186 , -0.7132575 , ...,  0.3606296 ,
           0.6993879 , -1.08

In [17]:
bert_results['sequence_output']

<tf.Tensor: shape=(5, 128, 768), dtype=float32, numpy=
array([[[ 0.19451576,  0.25141674,  0.19075057, ..., -0.24845093,
          0.3856855 ,  0.13291033],
        [-0.5947865 , -0.39420375,  0.2524569 , ..., -0.7694676 ,
          1.1564167 ,  0.32475808],
        [ 0.00641458, -0.15766454,  0.54610234, ..., -0.17451042,
          0.60289645,  0.42672288],
        ...,
        [ 0.23158628,  0.04302792,  0.7753697 , ...,  0.06123903,
          0.09081614, -0.14486143],
        [ 0.1565657 ,  0.03811888,  0.8351701 , ...,  0.05899143,
          0.0470343 , -0.09123509],
        [ 0.1898347 ,  0.03888722,  0.83606124, ...,  0.04247576,
          0.02342633, -0.10903205]],

       [[ 0.2705405 , -0.0782653 , -0.10499355, ..., -0.18821752,
          0.30700308,  0.0606564 ],
        [-0.34593925, -1.2121471 , -0.46464372, ..., -0.6203128 ,
          0.91636324, -0.16784051],
        [ 0.6963929 , -0.57969   ,  0.11662694, ..., -0.25387254,
          0.00638419, -0.61899734],
        ...,

Since we are using BERT base model it has 12 encoder layers, that's why above the length of encoder_output is 12. Read more about purpose of each <br>
element here: https://tfhub.dev/tensorflow/bert_en_uncased_L-12_H-768_A-12/4

You can see below that last element of encoder_outputs is basically a sequence_output

In [15]:
bert_results['encoder_outputs'][-1] == bert_results['sequence_output']

<tf.Tensor: shape=(5, 128, 768), dtype=bool, numpy=
array([[[ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        ...,
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True]],

       [[ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        ...,
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True]],

       [[ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        [ True,  True,  True, ...,  True,  True,  True],
        ...,
        [ True,  True,  True, ...,  True,  True,  